### DISCUSSION TITLE
Implement trailing slash convention

### DISCUSSION THREAD

**Discussion by cr101** [epicweb-dev/epic-stack#108] [Ideas]
Website crawlers consider `https://example.com/foo/` and `https://example.com/foo` as two separate URls.

In order to eliminate duplicate content issues and improve SEO, should epic-stack automatically redirect all slash URLs to the equivalent non-slash?
[How Trailing Slashes Affect Your Website’s SEO](https://authenticdigital.nz/blog/trailing-slashes-and-seo/)

**Comment by kentcdodds** [comment]
100%

I've just pushed this improvement. Just copied the implementation from my own site. Thanks!

**Reply by kiliman** [reply]
Hmm.. I believe that `index` routes should *always* end with `/`. This makes relative routes work correctly. Browsers base relative path up to nearest `/`.

You don't have to worry about SEO if you follow this rule. Granted, I understand that some hosts will strip those or don't work well with trailing `/`. I say, you should probably pick a new host.

Without `/`
```
https://example.com/users 
  <a href="kiliman">kiliman</a>
=> ‼️ https://example.com/kiliman
```
With `/` for index routes
```
https://example.com/users/
  <a href="kiliman">kiliman</a>
=> ✅ https://example.com/users/kiliman
```

**Reply by kentcdodds** [reply]
Regarding the relative `href`, I do not believe that's accurate. Could you make a reproduction?

**Reply by kiliman** [reply]
Here's a quick example

https://stackblitz.com/edit/stackblitz-starters-ghpc9d?description=&file=index.js&title=Express%20Starter

**Reply by kentcdodds** [reply]
Ah, yes, I was mistaken. Is this a problem in a Remix setting where you're using `<Link />` though? I don't think so 🤔

**Reply by kiliman** [reply]
With Remix and React Router, relative paths are route-based by default, not path-based (you can use `relative="path"`). That's why `<Link to="kiliman">` works since it knows you're on the `users` route already.

I only point it out since you're teaching _web development_ and not just Remix/RR.

**Reply by kentcdodds** [reply]
That's a fair point. I'll keep that in mind when I teach about relative routes 👍

The fact is that users do sometimes add/remove trailing slashes, so it's better to redirect from one to the other. Most sites default to no trailing slash (maybe because they look gross... at least that's what I think) so that's what we're going to do.

**Reply by richardscarrott** [reply]
I noticed Epic Stack redirects [POST requests with trailing slash](https://github.com/epicweb-dev/epic-stack/blob/04aececc5d3137713f9f01d132f3a25a122aff78/server/index.ts#L64).

Given it's a 301 (and not 308), I think it'll go from POST -> GET which probably isn't good? I'm guessing for SEO we only really need to worry about GET requests?

**Reply by kentcdodds** [reply]
I've not experienced an issue, but I agree it's probably wrong to redirect anything but GET requests. I'll make that change now.

Also, I highly discourage using 308. You never know what the future holds!

**Reply by richardscarrott** [reply]
Does a 301 not have the same caching risks as 308?...maybe it should be a 302?

**Reply by kentcdodds** [reply]
🤦‍♂️ You're right, this should be a 302. Fixed.

**Comment by zwhitchcox** [comment]
~kentcdodds~ curious why only redirect if the `req.path.length > 1`? This means that the home page trailing slash won't be directed. Is that intentional?

**Reply by zwhitchcox** [reply]
Nevermind:

>The good news is that it's not too difficult to fix issues with trailing slashes, depending on your website platform. It's important to note that root domain (often the homepage ‐ in our case [https://www.onlineassetpartners.co.nz](https://authenticdigital.nz/)) will always resolve without a trailing slash. Google does this automatically, and you won't be able to change it.
